pkg/kclient/deployments.go

```diff
@@ -225,20 +226,78 @@ func (c *Client) UpdateDeployment(deploy appsv1.Deployment) (*appsv1.Deployment,
 // odo overrides it with the value it expects instead of failing due to conflict.
 func (c *Client) ApplyDeployment(deploy appsv1.Deployment) (*appsv1.Deployment, error) {
 	data, err := json.Marshal(deploy)
-
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to marshal deployment")
+	}
 	klog.V(5).Infoln("Applying Deployment via server-side apply:")
 	klog.V(5).Infoln(resourceAsJson(deploy))
 
+	err = c.removeDuplicateEnv(deploy.Name)
 	if err != nil {
-		return nil, errors.Wrapf(err, "unable to marshal deployment")
+		return nil, err
 	}
+
 	deployment, err := c.KubeClient.AppsV1().Deployments(c.Namespace).Patch(context.TODO(), deploy.Name, types.ApplyPatchType, data, metav1.PatchOptions{FieldManager: FieldManager, Force: boolPtr(true)})
 	if err != nil {
 		return nil, errors.Wrapf(err, "unable to update Deployment %s", deploy.Name)
 	}
 	return deployment, nil
 }
 
+// removeDuplicateEnv removes duplicate environment variables from containers, due to a bug in Service Binding Operator:
+// https://github.com/redhat-developer/service-binding-operator/issues/983
+func (c *Client) removeDuplicateEnv(deploymentName string) error {
+	deployment, err := c.KubeClient.AppsV1().Deployments(c.Namespace).Get(context.Background(), deploymentName, metav1.GetOptions{})
+	if kerrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	changes := false
+	containers := deployment.Spec.Template.Spec.Containers
+	for i, container := range containers {
+		duplicates := []corev1.EnvVar{}
+		found := map[string]bool{}
+		envs := container.Env
+		for _, env := range envs {
+			if _, ok := found[env.Name]; ok {
+				duplicates = append(duplicates, env)
+			} else {
+				found[env.Name] = true
+			}
+		}
+		if len(duplicates) > 0 {
+			changes = true
+			newEnv := []corev1.EnvVar{}
+			// first remove duplicates
+			for _, env := range envs {
+				found := false
+				for _, duplicate := range duplicates {
+					if env.Name == duplicate.Name {
+						found = true
+						break
+					}
+				}
+				if !found {
+					newEnv = append(newEnv, env)
+				}
+			}
+			// next add them as single values
+			newEnv = append(newEnv, duplicates...)
+			containers[i].Env = newEnv
+		}
+	}
```

Comment:
I think we can simplify this to
```
for i := range containers {
	found := map[string]bool{}
	var newEnv []corev1.EnvVar
	for _, env := range containers[i].Env {
		if _, ok := found[env.Name]; !ok {
			found[env.Name] = true
			newEnv = append(newEnv, env)
		} else {
			changes = true
		}
	}
	containers[i].Env = newEnv
}
```
Am I missing something?

Reply:
Thanks :+1: